### PR TITLE
Uplift request form

### PR DESCRIPTION
**Pull request by La0**
Needs mozilla-conduit/lando-api#80

This PR adds a new form to fill create an uplift request, using the URL `/uplift/DXXX`.

The user needs to select at least one uplift repository (with `approval_required`) and fill most of the fields displayed.

Once the uplift is created, the list of new revisions is displayed, with direct links.

Here are some screenshots for the uplift form & the success stage after an uplift is created:

### CODE REVIEW

**Comment by La0** [comment]
Also updated to support both uplift & approval workflows.

File: landoui/forms.py [hunk 1]
```diff
@@ -90,3 +93,83 @@ class UserSettingsForm(FlaskForm):
...
+    A simple boolean field displayed as a list of radio buttons
+    Will output a bool value (or None on unknown)
+    """
+
+    def post_validate(self, *args, **kwargs):
+        values = {'yes': True, 'no': False, 'unknown': None}
+        self.data = values.get(self.data)
+
+
+class ApprovalRequestForm(FlaskForm):
```
Comment (mars-f) [inline landoui/forms.py]: Can we rename this to something uplift-specific, such as `UpliftApprovalForm`?  There is also a sec-approval request form and the generic name `ApprovalRequestForm` doesn't specify which workflow `lando.forms.ApprovalRequestForm` is used in.

File: landoui/forms.py [hunk 2]
```diff
@@ -90,3 +93,83 @@ class UserSettingsForm(FlaskForm):
...
+    """Form to create a new approval request"""
```
Comment (mars-f) [inline landoui/forms.py]: Likewise, this docstring should be uplift-specific.  "approval request" doesn't specify which of the different kinds of lando approval request this class works with.

File: landoui/uplift.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        repositories=stack["uplift_repositories"],
+        form=form,
+        uplift_request=uplift_request,
+
+        # Display a warning when phabricator is not available
+        phabricator_token_available=api.has_phabricator_token(),
+    )
+
+
+@uplift.route('/approval/D<int:revision_id>/', methods=('GET', 'POST'))
```
Comment (mars-f) [inline landoui/uplift.py]: We may want to name this endpoint to clarify that it's for uplift approval, not sec-approval.

File: landoui/templates/uplift/approval.html [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      </article>
+
+      {% elif repository.approval_required %}
+
+        {% set form_view = "uplift.approval" %}
+        {% include "uplift/_form.html" %}
+
+      {% else %}
+        <div class="notification is-info">
+          The target repository <strong>{{ repository.short_name }}</strong> does not need an approval. You can land directly the patch.
```
Comment (mars-f) [inline landoui/templates/uplift/approval.html]: I think you meant to say "land the patch directly to it".

**Comment by La0** [comment]
I just updated the PR adressing @mars-f comments. Could you take another look please ?

Sorry for the long delay for that iteration, i was busy on code-review :cry:

File: landoui/templates/stack/stack.html [hunk 5]
```diff
@@ -28,6 +28,9 @@ <h1>Stack containing revision {{revisions[revision_phid]['id']}}</h1>
           <th>Status</th>
           <th>Revision</th>
           <th>Reviewers</th>
+          {% if target_repo.approval_required %}
+          <th>Approval</th>
+          {% endif %}
```
Comment (smacleod) [inline landoui/templates/stack/stack.html]: the `target_repo` is for the landable series which may just be a small piece of the whole stack. Throughout the stack the various revisions might have distinct repositories. How do things look under this scenario? It'd be great to see some screenshots with various types of stacks.

**Comment by La0** [comment]
I added [some screenshots on imgur](https://imgur.com/a/ax2sd1K) (easier to update than github) & moved the partials in a sub dir.

@smacleod I tried to add a revision on the uplift-target repo, but the `/stacks` endpoint does not give any landable path on that revision: https://phabricator-dev.allizom.org/D1642
Is that because there is no existing commit on that commit ? Could you add some history on that repo ?

Regarding the stack management, i was under the impression the uplift form would only handle one revision at a time for now ?

**Comment by La0** [comment]
@smacleod Could we further discuss this PR ? I can split it if you prefer to handle smaller changesets.

**Comment by zzzeid** [comment]
Closing PR until further notice.

**Pull request closed without merging**